File: rb101/rb101-rb109-small_problems/easy7/staggered_case2.rb
def staggered_case(string)
  result = string.chars
  need_upper = true
  result.each do |char|
    need_upper ? char.upcase! : char.downcase!
    need_upper = !need_upper if char =~ /[a-z]/i
  end
  result.join
end

p staggered_case('I Love Launch School!') == 'I lOvE lAuNcH sChOoL!'
p staggered_case('ALL CAPS') == 'AlL cApS'
p staggered_case('ignore 77 the 444 numbers') == 'IgNoRe 77 ThE 444 nUmBeRs'

# ALTERNATIVE SOLUTION:

# def staggered_case(string)
#   result = ''
#   need_upper = true
#   string.chars.each do |char|
#     if char =~ /[a-z]/i
#       if need_upper
#         result += char.upcase
#       else
#         result += char.downcase
#       end
#       need_upper = !need_upper
#     else
#       result += char
#     end
#   end
#   result
# end

# FURTHER EXPLORATION:

# def true_or_false?(choice)
#   case choice
#   when true  then true
#   when false then false
#   else
#     raise 'Keyword argument must be a boolean value (true or false)!'
#   end
# end

# def staggered_case(string, count_non_alpha:)
#   result = string.chars
#   true_or_false?(count_non_alpha)
#   need_upper = true
#   result.each do |char|
#     need_upper ? char.upcase! : char.downcase!
#     if count_non_alpha
#       need_upper = !need_upper if char =~ /[a-z]/i
#     else
#       need_upper = !need_upper
#     end
#   end
#   result.join
# end

# p staggered_case('ALL CAPS', count_non_alpha: true) == 'AlL cApS' # => true
# p staggered_case('ALL CAPS', count_non_alpha: false) == 'AlL CaPs' # => true
# p staggered_case('ALL CAPS', count_non_alpha: 'true') == 'AlL cApS' # => Exception raised.

File: js210/js210-small_problems/string_and_text_processing/staggeredCase2.js
function staggeredCase(string) {
  let needUpper = true;

  return string.split('').map((char) =>  {
    if (/[a-z]/i.test(char)) {
      if (needUpper) {
        needUpper = false;
        return char.toUpperCase();
      } else {
        needUpper = true;
        return char.toLowerCase();
      }
    } else {
      return char;
    }
  }).join('');
}

console.log(staggeredCase('I Love Launch School!'));        // "I lOvE lAuNcH sChOoL!"
console.log(staggeredCase('ALL CAPS'));                     // "AlL cApS"
console.log(staggeredCase('ignore 77 the 444 numbers'));    // "IgNoRe 77 ThE 444 nUmBeRs"


// ALTERNATIVE SOLUTIONS:

// 1) - Launch School solution:
// function staggeredCase(string) {
//   let needUpper = true;
//   let newChar;

//   return string.split('').map(char => {
//     if (char.match(/[a-z]/i)) {
//       if (needUpper) {
//         newChar = char.toUpperCase();
//       } else {
//         newChar = char.toLowerCase();
//       }

//       needUpper = !needUpper;
//       return newChar;
//     } else {
//       return char;
//     }
//   }).join('');
// }

// 2) - Interesting solution from a classmate exploring the functionality of the String.prototype.replace method:
// function staggeredCase(string) {
//   return string.replace(/([a-z])([^a-z]*)([a-z])/gi, (_match, p1, p2, p3) => {
//     return p1.toUpperCase() + p2 + p3.toLowerCase();
//   });
// }
